osu.Game/Graphics/UserInterface/LoadingButton.cs

```diff
@@ -0,0 +1,97 @@
+﻿// Copyright (c) ppy Pty Ltd <contact@ppy.sh>. Licensed under the MIT Licence.
+// See the LICENCE file in the repository root for full licence text.
+
+using osu.Framework.Graphics;
+using osu.Framework.Graphics.Containers;
+using osu.Framework.Input.Events;
+using osu.Game.Graphics.Containers;
+using osuTK;
+
+namespace osu.Game.Graphics.UserInterface
+{
+    public abstract class LoadingButton : OsuHoverContainer
+    {
+        private const float fade_duration = 200;
+
+        private bool isLoading;
+
+        public bool IsLoading
+        {
+            get => isLoading;
+            set
+            {
+                isLoading = value;
+
+                Enabled.Value = !isLoading;
+
+                if (value)
+                {
+                    loading.Show();
+                    text.FadeOut(fade_duration, Easing.OutQuint);
+                    OnLoadingStart();
+                }
+                else
+                {
+                    loading.Hide();
+                    text.FadeIn(fade_duration, Easing.OutQuint);
+                    OnLoadingFinished();
+                }
+            }
+        }
+
+        public Vector2 LoadingAnimationSize
+        {
+            get => loading.Size;
+            set => loading.Size = value;
+        }
+
+        private readonly LoadingAnimation loading;
+        private readonly Drawable text;
+
+        protected LoadingButton()
+        {
+            Container content;
+
+            Child = content = CreateContent();
+
+            content.AddRange(new[]
+            {
+                text = CreateText(),
+                loading = new LoadingAnimation
+                {
+                    Anchor = Anchor.Centre,
+                    Origin = Anchor.Centre,
+                    Size = new Vector2(12)
+                }
+            });
+        }
+
+        protected override bool OnClick(ClickEvent e)
+        {
+            if (!Enabled.Value)
+                return false;
+
+            try
+            {
+                return base.OnClick(e);
+            }
+            finally
+            {
+                // run afterwards as this will disable this button.
+                IsLoading = true;
+            }
+        }
+
+        protected virtual void OnLoadingStart()
+        {
+        }
+
+        protected virtual void OnLoadingFinished()
+        {
+        }
+
+        protected abstract Container CreateContent();
```

Comment:
I don't feel good with this returning a container which has things added to it. It's not what I'd expect from "Create**Content**".

It'd be nice to have `CreateContent()` be a final thing, returning a non-descript `Drawable`. This could be accomplished by removing `CreateText()` and letting consumers decide what to hide inside `OnLoading` / `OnLoaded`.